### DISCUSSION TITLE
New release 1.5.1

### DISCUSSION THREAD

**Discussion by pablosnt** [pablosnt/rekono#190] [Announcements]
### Fixed

- Error during `executions-worker` build in Docker (https://github.com/pablosnt/rekono/issues/184)
- Error during database migration in Docker (https://github.com/pablosnt/rekono/issues/183)

<hr /><em>This discussion was created from the release <a href='https://github.com/pablosnt/rekono/releases/tag/1.5.1'>1.5.1</a>.</em>

**Comment by LasseKrache** [comment]
Hi Pablo,
After trying to install your tool via docker again and again on different systems (kali and Mac), I finally get it running with the DMG file! Thank you so much for that!
Unfortunately I cannot logon. :-(
Tried "rekono" as username & password, but no success. Can you help please?

Greetings and thanx again for your project!
Lasse

**Reply by pablosnt** [reply]
Thank you very much for using Rekono. 

Rekono Desktop installed via DMG provided in the release, needs the Rekono backend on https://127.0.0.1 (Docker). Did you try Desktop app in combination with Docker version? Note that Rekono Desktop is just another client of the Rekono backend, not a complete standalone app.

About Docker errors, did you try the version 1.5.1 that I release a few minutes ago? This version fixes some errors in Docker environment.

Let me know if I can help you in something else. Thanks!

**Comment by LasseKrache** [comment]
Ah, okay, didn't know that. :-O
Unfortunately the installation via docker failed. The error was:

_=> CANCELED [11/27] RUN apt install nmap dirsearch theharvester nikto sslscan sslyze cmseek zaproxy exploitdb metasp  97.4s
failed to solve: executor failed running [/bin/sh -c npm run build]: exit code: 1_

No idea, what I can change to solve that. :-(

**Reply by pablosnt** [reply]
Oh, other user had reported something like this today but I couldn't reproduce it. Could you execute `docker-compose build frontend` and send me the output? With this command, the error message will be more detailed. Moreover, just to know, could you tell me what docker-compose version are you using? (`docker-compose --version`). Thank you very much in advance!

**Reply by LasseKrache** [reply]
I hope this helps?

	▶ docker-compose --version
`Docker Compose version 2.16.0`

Started the run with

	▶ sudo docker-compose build frontend

```

...
 > [ 9/11] RUN npm install:
#0 14.26 npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
#0 15.29 npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
#0 15.48 npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
#0 15.51 npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
#0 15.74 npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
#0 15.97 npm WARN deprecated sourcemap-codec@1.4.8: Please use @jridgewell/sourcemap-codec instead
#0 16.52 npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
#0 17.67 npm WARN deprecated querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
#0 18.77 npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
#0 20.39 npm WARN deprecated har-validator@5.1.5: this library is no longer supported
#0 21.05 npm WARN deprecated html-webpack-plugin@3.2.0: 3.x is no longer supported
#0 21.33 npm WARN deprecated eslint-loader@2.2.1: This loader has been deprecated. Please use eslint-webpack-plugin
#0 23.53 npm WARN deprecated electron-osx-sign@0.6.0: Please use @electron/osx-sign moving forward. Be aware the API is slightly different
#0 24.47 npm WARN deprecated svgo@1.3.2: This SVGO version is no longer supported. Upgrade to v2.x.x.
#0 24.65 npm WARN deprecated babel-eslint@10.1.0: babel-eslint is now @babel/eslint-parser. This package will no longer receive updates.
#0 24.68 npm WARN deprecated asar@3.2.0: Please use @electron/asar moving forward.  There is no API change, just a package name change
#0 25.76 npm WARN deprecated @npmcli/move-file@1.1.2: This functionality has been moved to @npmcli/fs
#0 25.92 npm WARN deprecated @hapi/topo@3.1.6: This version has been deprecated and is no longer supported or maintained
#0 26.08 npm WARN deprecated @hapi/bourne@1.3.2: This version has been deprecated and is no longer supported or maintained
#0 26.75 npm WARN deprecated @hapi/address@2.1.4: Moved to 'npm install @sideway/address'
#0 27.53 npm WARN deprecated @hapi/hoek@8.5.1: This version has been deprecated and is no longer supported or maintained
#0 29.30 npm WARN deprecated chokidar@2.1.8: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies
#0 29.65 npm WARN deprecated chokidar@2.1.8: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies
#0 32.53 npm WARN deprecated @hapi/joi@15.1.1: Switch to 'npm install joi'
#0 34.62 npm WARN deprecated popper.js@1.16.1: You can find the new Popper v2 at @popperjs/core, this package is dedicated to the legacy v1
#0 94.56 npm notice
#0 94.56 npm notice New minor version of npm available! 9.4.0 -> 9.6.1
#0 94.56 npm notice Changelog: <https://github.com/npm/cli/releases/tag/v9.6.1>
#0 94.56 npm notice Run `npm install -g npm@9.6.1` to update!
#0 94.56 npm notice
#0 94.56 npm ERR! code 1
#0 94.56 npm ERR! path /code/node_modules/electron
#0 94.57 npm ERR! command failed
#0 94.57 npm ERR! command sh -c node install.js
#0 94.57 npm ERR! Error: ENOSPC: no space left on device, mkdir '/code/node_modules/electron/dist/resources'
#0 94.57
#0 94.57 npm ERR! A complete log of this run can be found in:
#0 94.57 npm ERR!     /home/rekono/.npm/_logs/2023-03-10T20_53_31_025Z-debug-0.log
------
failed to solve: executor failed running [/bin/sh -c npm install]: exit code: 1                                       /2m0,2s



```

**Reply by LasseKrache** [reply]
And this is the error I get on my kali linux:

```
└─$ sudo docker-compose up -d --scale executions-worker=5
Pulling initialize (rekono-kali:latest)...
ERROR: The image for the service you're trying to recreate has been removed. If you continue, volume data could be lost. Consider backing up your data before continuing.

Continue with the new image? [yN]y
Pulling initialize (rekono-kali:latest)...
ERROR: pull access denied for rekono-kali, repository does not exist or may require 'docker login': denied: requested access to the resource is denied

```

I will not mix this issues ... it's just to inform you. ;-)

**Reply by pablosnt** [reply]
Thank you very much for the details.

The error in the `frontend` build on MacOS seems a problem with your Docker storage space. With a fast google search, I have found this: https://stackoverflow.com/questions/50142049/enospc-no-space-left-on-device-nodejs. Can you try it?

The error in Kali Linux seems that you didn't execute the `docker-compose build` before `docker-compose up -d` command. The `build` command is the one who creates the Rekono images. Please confirm me if this is right or not.

I hope this to be useful for you, let me know if it works :)

**Reply by LasseKrache** [reply]
Thank you so much for your help!
I will take a look at the Mac solution tomorrow - it's after 10pm here in Germany and after a long working week I am just to tired to get into it. 
Maybe thats also the reason for the missing command in Kali. ;-)

I will report here again! Buen fin de semana!

**Reply by pablosnt** [reply]
You are welcome!

**Comment by LasseKrache** [comment]
Dear Pablo,

Today I started again with the docker installation on my Kali machine. First everything seems to run as expected, but in the end it seems that the nginx is rebooting again and again. See the attached log:


```
┌──(lk@pc1111)-[~/rekono]
└─$ sudo docker-compose build frontend 
[sudo] Passwort für lk: 
Building frontend
Sending build context to Docker daemon 10.04MB
Step 1/13 : FROM node:19.6.1-alpine
---> 4615c4abcbab
Step 2/13 : RUN mkdir /code
---> Using cache
---> 892e59c2d0ce
Step 3/13 : COPY /rekono/frontend /code
---> Using cache
---> 15a2149811b6
Step 4/13 : RUN apk update && apk add curl
---> Using cache
---> aca604892d25
Step 5/13 : ENV NODE_OPTIONS --openssl-legacy-provider
---> Using cache
---> 40db80a4719d
Step 6/13 : WORKDIR /code
---> Using cache
---> 2e2de3e24f3b
Step 7/13 : RUN npm install -g serve
---> Using cache
---> 349828e930b0
Step 8/13 : RUN adduser --disabled-password rekono
---> Using cache
---> 4b350f022f43
Step 9/13 : RUN chown -R rekono /code
---> Using cache
---> 7de1e2c25eb0
Step 10/13 : USER rekono
---> Using cache
---> 8757d967d0a8
Step 11/13 : RUN npm install
---> Using cache
---> 2d5e854709fb
Step 12/13 : RUN npx browserslist@latest --update-db
---> Using cache
---> f85f9eb6c6b2
Step 13/13 : RUN npm run build
---> Using cache
---> e9a6472939f7
Successfully built e9a6472939f7
Successfully tagged rekono-frontend:latest


# executing "sudo docker-compose up -d --scale executions-worker=5" without errors 
...
# after that I was unable to connect to 127.0.0.1 so I checked the containers:

┌──(lk@pc1111)-[~/rekono]
└─$ sudo docker container ls
CONTAINER ID IMAGE COMMAND CREATED STATUS PORTS NAMES
4fcaac24f1d8 rekono-nginx:latest "/entrypoint.sh" 2 hours ago Up Less than a second 0.0.0.0:80->80/tcp, :::80->80/tcp, 0.0.0.0:443->443/tcp, :::443->443/tcp nginx
21eec4bfac34 rekono-backend:latest "python manage.py te…" 2 hours ago Up 2 hours telegram-bot
567f534ef0fc rekono-kali:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_executions-worker_1
4ddcbd2a762b rekono-kali:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_executions-worker_3
c16207b30ec9 rekono-kali:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_executions-worker_4
4534ffcc9fbc rekono-backend:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_findings-worker_1
30e39cf7f606 rekono-kali:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_executions-worker_2
c63f282cb57d rekono-backend:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_emails-worker_1
f8beda032f09 rekono-kali:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_executions-worker_5
1ccdd50e3b1d rekono-backend:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_tasks-worker_1
dc51f5d8b198 rekono-backend:latest "python manage.py ru…" 2 hours ago Up 2 hours (healthy) 8000/tcp backend
9e3914e722a4 rekono-frontend:latest "docker-entrypoint.s…" 2 hours ago Up 2 hours (healthy) frontend
9fb73ecae096 postgres:14.2-alpine "docker-entrypoint.s…" 18 hours ago Up 3 hours (healthy) postgres
7737bfc951a7 redis:6.2.6-alpine "docker-entrypoint.s…" 18 hours ago Up 3 hours redis


┌──(lk@pc1111)-[~/rekono]
└─$ sudo docker container ls
CONTAINER ID IMAGE COMMAND CREATED STATUS PORTS NAMES
4fcaac24f1d8 rekono-nginx:latest "/entrypoint.sh" 2 hours ago Restarting (1) 43 seconds ago nginx
21eec4bfac34 rekono-backend:latest "python manage.py te…" 2 hours ago Up 2 hours telegram-bot
567f534ef0fc rekono-kali:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_executions-worker_1
4ddcbd2a762b rekono-kali:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_executions-worker_3
c16207b30ec9 rekono-kali:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_executions-worker_4
4534ffcc9fbc rekono-backend:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_findings-worker_1
30e39cf7f606 rekono-kali:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_executions-worker_2
c63f282cb57d rekono-backend:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_emails-worker_1
f8beda032f09 rekono-kali:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_executions-worker_5
1ccdd50e3b1d rekono-backend:latest "python manage.py rq…" 2 hours ago Up 2 hours rekono_tasks-worker_1
dc51f5d8b198 rekono-backend:latest "python manage.py ru…" 2 hours ago Up 2 hours (healthy) 8000/tcp backend
9e3914e722a4 rekono-frontend:latest "docker-entrypoint.s…" 2 hours ago Up 2 hours (healthy) frontend
9fb73ecae096 postgres:14.2-alpine "docker-entrypoint.s…" 18 hours ago Up 3 hours (healthy) postgres
7737bfc951a7 redis:6.2.6-alpine "docker-entrypoint.s…" 18 hours ago Up 3 hours redis
```

As you can see: the build has no problems, same for "docker up". But then I see the nginx container is rebooting every minute ... again and again.

Maybe (hopefully) this is just a newbie problem, which can be solved in seconds - I am grateful for any advice!

**Comment by LasseKrache** [comment]
And a little update from the docker installation on my Mac:
Same error! It seems that the nginx container is rebooting again and again, too!
Probably it's just my aura. ;-)

**Reply by pablosnt** [reply]
Mmm, it's strange. Do you see some error in logs? `docker-compose logs nginx`. I've just tried it again on Mac and it is working for me. I'm sorry for asking you to do or review things over and over again :(

**Reply by LasseKrache** [reply]
I can see this:

```
nginx  | /docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
nginx  | /docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
nginx  | /docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
nginx  | 10-listen-on-ipv6-by-default.sh: info: IPv6 listen already enabled
nginx  | /docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
nginx  | /docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
nginx  | /docker-entrypoint.sh: Configuration complete; ready for start up
nginx  | 2023/03/11 18:26:11 [emerg] 1#1: cannot load certificate "/etc/nginx/tls/certificate.crt": BIO_new_file() failed (SSL: error:02001002:system library:fopen:No such file or directory:fopen('/etc/nginx/tls/certificate.crt','r') error:2006D080:BIO routines:BIO_new_file:no such file)
nginx  | nginx: [emerg] cannot load certificate "/etc/nginx/tls/certificate.crt": BIO_new_file() failed (SSL: error:02001002:system library:fopen:No such file or directory:fopen('/etc/nginx/tls/certificate.crt','r') error:2006D080:BIO routines:BIO_new_file:no such file)
```

**Reply by LasseKrache** [reply]
In Kali a certificate problem seems to be the reason, too:

```
nginx                | writing new private key to '/etc/nginx/tls/privatekey.key'
nginx                | req: Can't open "/etc/nginx/tls/privatekey.key" for writing, Permission denied
nginx                | Generating a RSA private key
nginx                | ...........................................................++++
nginx                | .....................................++++
nginx                | writing new private key to '/etc/nginx/tls/privatekey.key'
nginx                | req: Can't open "/etc/nginx/tls/privatekey.key" for writing, Permission denied
nginx                | Generating a RSA private key
nginx                | ................................++++
nginx                | ...........++++
nginx                | writing new private key to '/etc/nginx/tls/privatekey.key'
nginx                | req: Can't open "/etc/nginx/tls/privatekey.key" for writing, Permission denied
```

**Reply by pablosnt** [reply]
I don't know why occurs in your case, but the error means that the TLS certificate and key are not generated automatically. Just to let you know, when you don't configure your own TLS certificate to be used, an self-signed certificate is generated automatically here: https://github.com/pablosnt/rekono/blob/main/docker/nginx/entrypoint.sh#L5. As I said, I don't know why it doesn't happen in your case, but I think that the error can be fixed if you execute the following command from the root directory of Rekono project:

```
openssl req -x509 \
        -nodes \
        -days 365 \
        -newkey rsa:4096 \
        -sha512 \
        -keyout docker/nginx/tls/privatekey.key \
        -out docker/nginx/tls/certificate.crt \
        -subj "/C=ES/ST=Spain/L=Spain/O=Rekono/OU=Security/CN=Rekono"
```

This will configure a self-signed certificate that the Nginx needs to use HTTPS to expose Rekono services. After that, when the Nginx container reboots if should work fine. Please, let me know if it works

**Reply by pablosnt** [reply]
Okay, I'm sorry, I didn't see your last comment. Did you clone Rekono repository as root user?

**Reply by pablosnt** [reply]
If you do, please retry to clone it with a non-root user or change the ownership of directory `docker/nginx/tls/` to your regular system user. This directory is mapped to the `/etc/nginx/tls/` container directory where the TLS certificate and key are generated, and I think that this could be the problem

**Reply by LasseKrache** [reply]
Dear Pablo,

Just to start with the import issue:
**Thank you! Thank you very, very much!**
I really appreciate all the effort you make for this project and for helping newbies like me! Encantado, de verdad!

On my Mac Rekono is working now - not via App, but via browser access, which is enough for me!
On Kali it seems, that I have to start from the beginning again: Although I cloned the repository with sudo and also created the required certificate, nginx is still rebooting.

BUT: never mind! I am very happy to try it rekono on my Mac!
Thank you SO MUCH for your help. 

Si estas un dia en el surest de Alemania, me dices - nos vamos a tomar unas cervezitas! ;-)

**Reply by pablosnt** [reply]
Great news! You are welcome, the pleasure is mine!

For Rekono Destkop, if you want to try it, once you have Rekono running in Docker, it should connect to your local server in port 443 automatically. I mean, you should be able to login.

For the Kali, just to clarify, you should clone Rekono repository WITHOUT sudo. If you need it, you have to provide access to the non-root user that Docker runs with.

**Reply by LasseKrache** [reply]
A small update to my Kali installation:

After changing the file permissions in the `docker/nginx/tls/` directory according to the ones I see on my Mac, everything ist working now!

Otra vez: Muchas gracias y que pases un buen fin de semana!

**Reply by pablosnt** [reply]
Great! Igualmente ;)

**Comment by skilfoy** [comment]
I am also having an issue with this tool I'm running a fresh install of the latest Kali (2023.1) on bare metal. I used `git clone` (no sudo) to clone the repo. Then I used `sudo docker-compose build` to build the images. Then I used `sudo docker-compose up -d --scale executions-worker=5`. 

`└─$ sudo docker-compose up -d --scale executions-worker=5
Starting redis    ... done
Starting frontend ... done
Starting postgres ... done
Starting initialize ... done
Starting rekono_tasks-worker_1      ... done
Starting rekono_findings-worker_1   ... done
Starting rekono_executions-worker_1 ... done
Starting rekono_emails-worker_1     ... done
Starting telegram-bot               ... done
Starting backend                    ... done
Creating rekono_executions-worker_2 ... done
Creating rekono_executions-worker_3 ... done
Creating rekono_executions-worker_4 ... done
Creating rekono_executions-worker_5 ... done
Starting nginx                      ... done
`

Everything looks good here but when I try navigating to the Rekono web interface, the browser says the page cannot be found. `Unable to connect. An error occurred during a connection to 127.0.0.1.` I tried running the openssl command posted elsewhere in this discussion thread, but that didn't work either, and I'm still getting the same errors.

`└─$ openssl req -x509 \                                    
        -nodes \
        -days 365 \
        -newkey rsa:4096 \
        -sha512 \
        -keyout docker/nginx/tls/privatekey.key \
        -out docker/nginx/tls/certificate.crt \
        -subj "/C=ES/ST=Spain/L=Spain/O=Rekono/OU=Security/CN=Rekono"
`

Here are the nginx logs:

`
└─$ sudo docker-compose logs nginx | tail -n5
nginx                | nginx: [emerg] cannot load certificate key "/etc/nginx/tls/privatekey.key": BIO_new_file() failed (SSL: error:0200100D:system library:fopen:Permission denied:fopen('/etc/nginx/tls/privatekey.key','r') error:2006D002:BIO routines:BIO_new_file:system lib)
nginx                | 2023/03/20 01:01:33 [emerg] 1#1: cannot load certificate key "/etc/nginx/tls/privatekey.key": BIO_new_file() failed (SSL: error:0200100D:system library:fopen:Permission denied:fopen('/etc/nginx/tls/privatekey.key','r') error:2006D002:BIO routines:BIO_new_file:system lib)
nginx                | nginx: [emerg] cannot load certificate key "/etc/nginx/tls/privatekey.key": BIO_new_file() failed (SSL: error:0200100D:system library:fopen:Permission denied:fopen('/etc/nginx/tls/privatekey.key','r') error:2006D002:BIO routines:BIO_new_file:system lib)
nginx                | 2023/03/20 01:02:34 [emerg] 1#1: cannot load certificate key "/etc/nginx/tls/privatekey.key": BIO_new_file() failed (SSL: error:0200100D:system library:fopen:Permission denied:fopen('/etc/nginx/tls/privatekey.key','r') error:2006D002:BIO routines:BIO_new_file:system lib)
nginx                | nginx: [emerg] cannot load certificate key "/etc/nginx/tls/privatekey.key": BIO_new_file() failed (SSL: error:0200100D:system library:fopen:Permission denied:fopen('/etc/nginx/tls/privatekey.key','r') error:2006D002:BIO routines:BIO_new_file:system lib)
`

I already made sure the `./docker/nginx/tls` folder (and everything else in the rekono directory are owned by my regular kali user. Just to make sure there were no problems with the `./docker/nginx/tls` directory, I ran `chmod 777` on it.

Lastly, here is the relevant part of the `docker ps` output:

`
└─$ sudo docker ps
CONTAINER ID   IMAGE                    COMMAND                  CREATED          STATUS                          PORTS                                       NAMES
3df2c7097256   rekono-nginx:latest      "/entrypoint.sh"         25 hours ago     Restarting (1) 23 seconds ago    nginx
`

Everything else in the `docker ps` output shows healthy / "up" statuses, except for this nginx frontend. As such, I am unable to use the program as intended.

Thanks in advance for all your help and I look forward to getting this running and putting this tool to work!

**Reply by pablosnt** [reply]
Hi @skilfoy,

I'm not very sure about the reason or the most proper way to fix this error. Anyway, I could reproduce the error so I can give you two options:

- Run `docker-compose` as non-root user. This is very difficult in Kali because you have to follow the following [guide](https://docs.docker.com/engine/security/rootless/) to install Docker, but you can try it if you want, I think that it could be the better approach.

- In my case, I try to clone Rekono repository as root (I never advise it) and execute both `docker-compose` commands also as root, and it works for me. It's important to execute before the `openssl` command (the same that you send before) and the command `chmod 777 -R docker/nginx/tls`. Of course, it isn't clean and it isn't a good practice, but it works.

Let me know if this options fix the error in your case or not.
Regards
